### PR TITLE
Implement PublicBuffer

### PR DESCRIPTION
**Pull request by ionut-arm**
This commit adds a new type to serve as a native placeholder for
TPM2B_PUBLIC, with Public being switched to represent TPMT_PUBLIC.

Conversions and Marshalling/UnMarshalling implementations are also added
for Public and PublicBuffer, along with tests for these conversions.

cc @rshearman

### CODE REVIEW

**Comment by Superhepper** [comment]
I need to try to read this without constantly thinking "structures needs to be refactored". But from a quick overview I think it looked pretty good.

**Comment by ionut-arm** [comment]
> I need to try to read this without constantly thinking "structures needs to be refactored". But from a quick overview I think it looked pretty good.

Yeah - as long as we don't change what users would see, we can do it in a future release.

Also, just a note: I've implemented the conversions to/from `TPM2B_PUBLIC` on `Public` as well because it's rather annoying otherwise - you have to go through marshalling/unmarshalling if you want to use `PublicBuffer` instead. So annoying to have a "buffer" type that isn't actually a buffer... 😐

File: tss-esapi/src/context/tpm_commands/object_commands.rs [hunk 1]
```diff
@@ -71,7 +71,7 @@ impl Context {
                 self.optional_session_2(),
                 self.optional_session_3(),
                 &sensitive_create,
-                &public.clone().into(),
+                &public.clone().try_into()?,
```
Comment (wiktor-k) [inline tss-esapi/src/context/tpm_commands/object_commands.rs:74]: This is not directly related to your PR but the `clone()` call... ugh... maybe we should track or decide on whether to take the Public by value (so that the caller has to explicitly clone if they want to reuse public). Because currently it always clones public even if passing by value a single thing would suffice...

File: tss-esapi/src/structures/buffers/public.rs [hunk 2]
```diff
@@ -1,507 +1,168 @@
...
-    /// [KeyedHash][`crate::interface_types::algorithm::PublicAlgorithm::KeyedHash].
-    pub fn with_keyed_hash_unique_identifier(
-        mut self,
-        keyed_hash_unique_identifier: &Digest,
-    ) -> Self {
-        self.keyed_hash_unique_identifier = Some(keyed_hash_unique_identifier.clone());
-        self
+    fn try_from(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() > Self::MAX_SIZE {
+            error!("Error: Invalid &[u8] size(> {})", Self::MAX_SIZE);
```
Comment (wiktor-k) [inline tss-esapi/src/structures/buffers/public.rs:60]: :thinking: why not attach `bytes.len()` to the error message such as:
```rust
error!("Error: Invalid &[u8] size: {} > {}", bytes.len(), Self::MAX_SIZE);
```

File: tss-esapi/src/structures/buffers/public.rs [hunk 3]
```diff
@@ -1,507 +1,168 @@
...
-    /// [Ecc][`crate::interface_types::algorithm::PublicAlgorithm::Ecc].
-    ///
-    /// The unique identifier is a ecc point.
-    pub fn with_ecc_unique_identifier(mut self, ecc_unique_identifier: &EccPoint) -> Self {
-        self.ecc_unique_identifier = Some(ecc_unique_identifier.clone());
-        self
+    fn try_from(tss: TPM2B_PUBLIC) -> Result<Self> {
+        let size = tss.size as usize;
+        if size > Self::MAX_SIZE {
+            error!("Error: Invalid buffer size(> {})", Self::MAX_SIZE);
```
Comment (wiktor-k) [inline tss-esapi/src/structures/buffers/public.rs:73]: Likewise.

File: tss-esapi/src/context/tpm_commands/object_commands.rs (hunk 1 above)
Comment (ionut-arm) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: So, I didn't notice this before in the Rust API guidelines, but it seems to actually be specified there: [we should ask for ownership instead of cloning from a reference](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control). Given that this essentially breaks the API, the question is whether we want to ship it in 7.0.0. Thoughts? Should we discuss it under #290 ? cc @Superhepper

Also, I'm not sure if I actually answered the question or if you meant that we could not take ownership at all, since we only need a reference in the end (though the reference is to `TPM2B_PUBLIC`)

Comment (ionut-arm) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: Damn, deleted my comment by accident...

It seems, from the Rust API guidelines, that [we should ask for ownership of the value instead of cloning from a reference](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control). Since this breaks the API we need to decide whether to do this for 7.0.0 or leave it for a subsequent release. It's annoying, but it's not the end of the world, and we have a lot of changes due in 7.0.0, so I'd prefer sticking with the current API, but can be convinced otherwise. Thoughts? cc @Superhepper 

Also, I don't know if the paragraph above answers your question - were you asking if we really need ownership here at all, since we only need to pass down a reference?

File: tss-esapi/src/structures/buffers/public.rs (hunk 2 above)
Comment (ionut-arm) [reply tss-esapi/src/structures/buffers/public.rs]: Fixed

File: tss-esapi/src/structures/buffers/public.rs (hunk 3 above)
Comment (ionut-arm) [reply tss-esapi/src/structures/buffers/public.rs]: Fixed

File: tss-esapi/src/context/tpm_commands/object_commands.rs (hunk 1 above)
Comment (wiktor-k) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: There are two things in play here:
  - first, conversion from `&Public` to `Public` that allows to omit the `clone()` call, that was what I was talking about, I don't think this is a big deal with regards to API breakage (I've been updating my crate continuously to 7-alpha and there were bigger issues but I understand if you'd like to defer that),
  - then I thought: it's actually just doing `try_into` to convert to `PublicBuffer` (checked docs for `TPM2B_PUBLIC`) so a next question raises: why not take `PublicBuffer` parameter instead?

Comment (Superhepper) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: We have already declared that there will be some huge interface changes in 7.0.0. And I think I would prefer that we do all the interface changes that we have been wanted to do now when we have the chance. I am already prepared that moving my code to 7.0.0 will be a huge task. And it would annoy me if I now fixed all my code and then some one makes interfaces changes for lets say 8.0.0 and new functionality that I need are based on those changes.

I am not to found of the method API dealing with PublicBuffer. Because, and correct me if I am wrong, that is not a type that is returned from the context methods. Public is the type we uses because it contains the values in a human readable form, PublicBuffer is something you use if you want the type as bytes.

Comment (ionut-arm) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: > it's actually just doing try_into to convert to `PublicBuffer`

It's not, that's why I implemented `TryFrom<Public> for TPM2B_PUBLIC`, so we don't have to go through `PublicBuffer`, because getting a `PublicBuffer` (or getting something from `PublicBuffer`) _always_ requires either marshalling or unmarshalling, so you'd have two useless operations in-between.

Comment (wiktor-k) [reply tss-esapi/src/context/tpm_commands/object_commands.rs:74]: Oh, great! I must've overlooked that fact. Thanks for the explanation!

**Review by wiktor-k — APPROVED** [review]

File: tss-esapi/src/structures/tagged/signature.rs [hunk 4]
```diff
@@ -125,3 +128,59 @@ impl TryFrom<TPMT_SIGNATURE> for Signature {
...
+                    error!("Failed to convert size of buffer to TSS size_t type: {}", e);
+                    Error::local_error(WrapperErrorKind::InvalidParam)
+                })?,
+                &mut offset,
+            )
+        };
+        let ret = Error::from_tss_rc(ret);
+        if ret.is_success() {
+            unsafe {
+                buffer.set_len(offset as usize);
```
Comment (Superhepper) [inline tss-esapi/src/structures/tagged/signature.rs:154]: This should be changed to the vector version that does not ned to use 'unsafe'.

**Review by Superhepper — COMMENTED** [review]
Looks good to me. Only had one remark. Don't know if that was in code written by you or something that have been moved but it would be nice to get it fixed.

File: tss-esapi/src/structures/tagged/signature.rs (hunk 4 above)
Comment (ionut-arm) [reply tss-esapi/src/structures/tagged/signature.rs:154]: Nice, thanks for pointing that out!

**Review by Superhepper — APPROVED** [review]
